fish-rust/src/env_dispatch.rs

```diff
@@ -0,0 +1,800 @@
+}
+
+fn handle_fish_history_change(vars: &dyn Environment) {
+    let fish_history = vars.get(L!("fish_history"));
+    let var = crate::env::env_var_to_ffi(fish_history);
+    crate::ffi::reader_change_history(&crate::ffi::history_session_id(var));
+}
+
+fn handle_fish_cursor_selection_mode_change(vars: &dyn Environment) {
+    use crate::reader::CursorSelectionMode;
+
+    let inclusive = vars
+        .get(L!("fish_cursor_selection_mode"))
+        .as_ref()
+        .map(|v| v.as_string())
+        .map(|v| v == L!("inclusive"))
+        .unwrap_or(false);
+    let mode = if inclusive {
+        CursorSelectionMode::Inclusive
+    } else {
+        CursorSelectionMode::Exclusive
+    };
+
+    let mode = mode as u8;
+    crate::ffi::reader_change_cursor_selection_mode(mode);
+}
+
+fn handle_autosuggestion_change(vars: &dyn Environment) {
+    // TODO: This was a call to reader_set_autosuggestion_enabled(vars) and
+    // reader::check_autosuggestion_enabled() should be private to the `reader` module.
+    crate::ffi::reader_set_autosuggestion_enabled_ffi(crate::reader::check_autosuggestion_enabled(
+        vars,
+    ));
+}
+
+fn handle_function_path_change(_: &dyn Environment) {
+    crate::ffi::function_invalidate_path();
+}
+
+fn handle_complete_path_change(_: &dyn Environment) {
+    crate::ffi::complete_invalidate_path();
+}
+
+fn handle_tz_change(var_name: &wstr, vars: &dyn Environment) {
+    handle_timezone(var_name, vars);
+}
+
+fn handle_locale_change(vars: &dyn Environment) {
+    init_locale(vars);
+    // We need to re-guess emoji width because the locale might have change to a multibyte one.
+    guess_emoji_width(vars);
+}
+
+fn handle_curses_change(vars: &dyn Environment) {
+    guess_emoji_width(vars);
+    init_curses(vars);
+}
+
+fn handle_fish_use_posix_spawn_change(vars: &dyn Environment) {
+    // Note that if the variable is missing or empty we default to true (if allowed).
+    if !allow_use_posix_spawn() {
+        USE_POSIX_SPAWN.store(false, Ordering::Relaxed);
+    } else if let Some(var) = vars.get(L!("fish_use_posix_spawn")) {
+        let use_posix_spawn =
+            var.is_empty() || crate::wcstringutil::bool_from_string(&var.as_string());
+        USE_POSIX_SPAWN.store(use_posix_spawn, Ordering::Relaxed);
+    } else {
+        USE_POSIX_SPAWN.store(true, Ordering::Relaxed);
+    }
+}
+
+/// Allow the user to override the limits on how much data the `read` command will process. This is
+/// primarily intended for testing, but could also be used directly by users in special situations.
+fn handle_read_limit_change(vars: &dyn Environment) {
+    let read_byte_limit = vars
+        .get_unless_empty(L!("fish_read_limit"))
+        .map(|v| v.as_string())
+        .and_then(|v| match v.to_string().parse() {
+            Ok(v) => Some(v),
+            Err(_) => {
+                // XXX: In the C++ code, this warning wasn't behind an "is_interactive_session()"
+                // check.
```

Comment:
Not really sold on this - $fish_read_limit is valid even non-interactively.

Reply:
Yeah, that makes sense. All the other warnings are being suppressed in non-interactive mode so I thought this could join them.